File: src/com/wix/rt/cli/RTCliBuilder.java
package com.wix.rt.cli;

import com.intellij.execution.configurations.GeneralCommandLine;
import com.wix.nodejs.NodeRunner;
import org.jetbrains.annotations.NotNull;

/**
 * build cli command for RT
 * Created by idok on 11/27/14.
 */
final class RTCliBuilder {

    public static final String MODULES = "--modules";
    public static final String DRY_RUN = "--dry-run";
    public static final String FORCE = "--force";
    public static final String FORMAT = "--format";
    public static final String JSON = "json";
    public static final String LIST_TARGET_VERSION = "--list-target-version";
    public static final String V = "-v";

    private RTCliBuilder() {
    }

    @NotNull
    static GeneralCommandLine listVersions(@NotNull RTSettings settings) {
        GeneralCommandLine commandLine = createCommandLine(settings);
        commandLine.addParameter(LIST_TARGET_VERSION);
        commandLine.addParameter(FORMAT);
        commandLine.addParameter(JSON);
        return commandLine;
    }

    @NotNull
    static GeneralCommandLine version(@NotNull RTSettings settings) {
        GeneralCommandLine commandLine = createCommandLine(settings);
        commandLine.addParameter(V);
        return commandLine;
    }

    @NotNull
    static GeneralCommandLine createCommandLine(@NotNull RTSettings settings) {
        return NodeRunner.createCommandLine(settings.cwd, settings.node, settings.rtExecutablePath);
    }

    @NotNull
    static GeneralCommandLine createCommandLineBuild(@NotNull RTSettings settings) {
        GeneralCommandLine commandLine = createCommandLine(settings);
//        if (!new File(settings.cwd, settings.targetFile).exists()) {
//            throw new IllegalArgumentException("targetFile doesn't exist");
//        }
        commandLine.addParameter(settings.targetFile);
        commandLine.addParameter(MODULES);
        commandLine.addParameter(settings.modules);
        if (settings.dryRun) {
            commandLine.addParameter(DRY_RUN);
        }
        commandLine.addParameter(FORMAT);
        commandLine.addParameter(JSON);
        return commandLine;
    }

    @NotNull
    static GeneralCommandLine createCommandLineValidate(@NotNull RTSettings settings) {
        GeneralCommandLine commandLine = createCommandLine(settings);
//        if (!new File(settings.cwd, settings.targetFile).exists()) {
//            throw new IllegalArgumentException("targetFile doesn't exist");
//        }
        commandLine.addParameter(settings.targetFile);
        commandLine.addParameter(MODULES);
        commandLine.addParameter(settings.modules);
        commandLine.addParameter(DRY_RUN); //no generation
        commandLine.addParameter(FORCE); //validate with force
        commandLine.addParameter(FORMAT);
        commandLine.addParameter(JSON);
        return commandLine;
    }
}

File: src/com/wix/rtk/cli/RTCliBuilder.kt
package com.wix.rtk.cli

import com.intellij.execution.configurations.GeneralCommandLine
import com.intellij.openapi.util.SystemInfo
import com.wix.nodejs.NodeRunner
import com.wix.rt.build.VerifyMessage
import java.io.File
import java.util.*

/**
 * build cli command for RT
 * Created by idok on 11/27/14.
 */
internal object RTCliBuilder {
    val MODULES = "--modules"
    val DRY_RUN = "--dry-run"
    val FORCE = "--force"
    val FORMAT = "--format"
    val NATIVE = "--native"
    val JSON = "json"
    val LIST_TARGET_VERSION = "--list-target-version"
    val V = "-v"

    fun listVersions(settings: RTSettings): GeneralCommandLine =
            CLI(settings.cwd, settings.node, settings.rtExe)
                    .param(LIST_TARGET_VERSION)
                    .param(FORMAT, JSON)
                    .command

    fun version(settings: RTSettings): GeneralCommandLine = CLI(settings.cwd, settings.node, settings.rtExe).param(V).command

    fun createCommandLineBuild(settings: RTSettings): GeneralCommandLine =
            CLI(settings.cwd, settings.node, settings.rtExe)
                    .param(settings.targetFile)
                    .param(MODULES, settings.modules)
                    .flag(settings.dryRun, DRY_RUN)
                    .flag(settings.reactNative, NATIVE)
                    .param(FORMAT, JSON)
                    .command

    fun createCommandLineValidate(settings: RTSettings): GeneralCommandLine =
            CLI(settings.cwd, settings.node, settings.rtExe)
                    .param(settings.targetFile)
                    .param(MODULES, settings.modules)
                    .param(DRY_RUN) //no generation
                    .param(FORCE) //validate with force
                    .param(FORMAT, JSON)
                    .flag(settings.reactNative, NATIVE)
                    .command
}

class CLI(val cwd: String, val node: String, val exePath: String) {
    val command = NodeRunner.createCommandLine(cwd, node, exePath)

    fun param(name: String): CLI {
        command.addParameter(name)
        return this
    }

    fun param(name: String, value: String): CLI {
        command.addParameter(name)
        command.addParameter(value)
        return this
    }

    fun flag(value: Boolean, flagName: String): CLI {
        if (value) {
            command.addParameter(flagName)
        }
        return this
    }
}

class CLI2(val cwd: String, val exe: String) {
    val command = createCommandLine()

    fun param(name: String): CLI2 {
        command.addParameter(name)
        return this
    }

    fun param(name: String, value: String): CLI2 {
        command.addParameter(name)
        command.addParameter(value)
        return this
    }

    fun flag(value: Boolean, flagName: String): CLI2 {
        if (value) {
            command.addParameter(flagName)
        }
        return this
    }

    fun createCommandLine(): GeneralCommandLine {
        val commandLine = GeneralCommandLine()
        if (!File(cwd).exists() || !File(exe).exists()) {
            throw IllegalArgumentException("path doesn't exist")
        }
        commandLine.setWorkDirectory(cwd)
        //        if (SystemInfo.isWindows) {
        commandLine.exePath = exe
        //        } else {
        //            if (!File(node).exists()) {
        //                throw IllegalArgumentException("path doesn't exist")
        //            }
        //            commandLine.setExePath(node)
        //        }
        return commandLine
    }
}

/**
 * RT result
 * Created by idok on 8/25/14.
 */
class Result(val warns: List<VerifyMessage>, val errorOutput: String = "")

//class VerifyMessage(var level: String = "", var msg: String = "", var file: String = "", var line: Int = 0, var column: Int = 0,
//                    var index: Int = 0, var startOffset: Int = 0, var endOffset: Int = 0) {
//    companion object {
//        val WARN = "WARN"
//        val ERROR = "ERROR"
//        val INFO = "INFO"
//    }
//}
